src/lib/bootstrap_controller/bootstrap_controller.ml

```diff
@@ -0,0 +1,244 @@
+    let module Ps = Consensus.Mechanism.Protocol_state in
+    let candidate = External_transition.protocol_state transition in
+    if isn't_worth_getting_root t candidate time_received then Deferred.unit
+    else
+      let previous_state_hash = Ps.previous_state_hash candidate in
+      (* TODO: This may be an off by one *)
+      let input : Ancestor.Input.t =
+        { descendant= previous_state_hash
+        ; generations= length t.best_with_root.root }
+      in
+      let rec get_root () =
+        let%bind res = get_ancestor ~network:t.network input in
+        if
+          done_syncing_root t
+          || isn't_worth_getting_root t candidate time_received
+        then Deferred.unit
+        else
+          match res with
+          | Error e ->
+              Logger.error t.logger !"%{sexp:Error.t}" e ;
+              get_root ()
+          | Ok (ancestor, proof) -> (
+              let ancestor_length =
+                Ps.(Consensus_state.length (consensus_state ancestor))
+              in
+              match
+                Ancestor.Prover.verify_and_add t.ancestor_prover input
+                  (Ps.hash ancestor) proof ~ancestor_length
+              with
+              | Ok () ->
+                  t.best_with_root <- {state= candidate; root= ancestor} ;
+                  let candidate_body_hash = Ps.Body.hash (Ps.body candidate) in
+                  let candidate_hash =
+                    Protocol_state.hash ~hash_body:Fn.id
+                      {body= candidate_body_hash; previous_state_hash}
+                  in
+                  Ancestor.Prover.add t.ancestor_prover ~hash:candidate_hash
+                    ~prev_hash:previous_state_hash
+                    ~length:
+                      Ps.(Consensus_state.length (consensus_state candidate))
+                    ~body_hash:candidate_body_hash ;
+                  let ledger_hash =
+                    Consensus.Mechanism.(
+                      Protocol_state.blockchain_state ancestor
+                      |> Blockchain_state.ledger_hash
+                      |> Frozen_ledger_hash.to_ledger_hash)
+                  in
+                  ( match
+                      Syncable_ledger.new_goal t.syncable_ledger ledger_hash
+                    with
+                  | `Ignore -> ()
+                  | `Continue ->
+                      Ledger_hash.Table.set ledger_hash_table ~key:ledger_hash
+                        ~data:candidate_hash ) ;
+                  Deferred.unit
+              | Error e -> received_bad_proof t e ; get_root () )
+      in
+      get_root ()
+
+  let should_bootstrap root_state new_state =
+    let new_length = length new_state in
+    let root_length = length root_state in
+    new_length - root_length
+    > (2 * Transition_frontier.max_length) + Consensus.Mechanism.network_delay
+
+  let setup_bootstrap ~ledger_hash_table ~toggle_pipes ~frontier t
+      (transition, tm) =
+    toggle_pipes () ;
```

Comment:
I've found `toggling` when it does mutation to be a source of annoying bugs, can you instead have explicit `start ()` and `stop ()` functions instead